analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs

```diff
@@ -0,0 +1,82 @@
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    /// <summary>
+    /// Immutable builder that holds all parameters for rule verification.
+    /// </summary>
+    internal class VerifierBuilder
+    {
+        // All properties are (and should be) immutable.
+        public ImmutableArray<Func<DiagnosticAnalyzer>> Analyzers { get; init; } = ImmutableArray<Func<DiagnosticAnalyzer>>.Empty;
+        public ImmutableArray<string> Paths { get; init; } = ImmutableArray<string>.Empty;
+        public ImmutableArray<MetadataReference> References { get; init; } = ImmutableArray<MetadataReference>.Empty;
+        public ImmutableArray<ParseOptions> ParseOptions { get; init; } = ImmutableArray<ParseOptions>.Empty;
+
+        public VerifierBuilder() { }
+
+        private VerifierBuilder(VerifierBuilder original)
+        {
+            Paths = original.Paths;
+            References = original.References;
+            ParseOptions = original.ParseOptions;
+            Analyzers = original.Analyzers;
+        }
+
+        /// <summary>
+        /// This method solves complicated scenarios. Use 'new VerifierBuilder&lt;TAnalyzer&gt;()' for single analyzer cases with no rule parameters.
+        /// </summary>
+        public VerifierBuilder AddAnalyzer(Func<DiagnosticAnalyzer> createConfiguredAnalyzer) =>
+            new(this) { Analyzers = Analyzers.Append(createConfiguredAnalyzer).ToImmutableArray() };
+
+        public VerifierBuilder AddPaths(params string[] paths) =>
+            new(this) { Paths = Paths.Concat(paths).ToImmutableArray() };
+
+        public VerifierBuilder AddReferences(IEnumerable<MetadataReference> references) =>
+            new(this) { References = References.Concat(references).ToImmutableArray() };
+
+        public VerifierBuilder WithOptions(ImmutableArray<ParseOptions> parseOptions) =>
+            new(this) { ParseOptions = parseOptions };
+
+        public Verifier Build() =>
+            new(this);
```

Comment:
The `Verifier` (the class which needs to be created) should not know about the `Builder` (there should be no dependency in this direction). Only the `Builder` should know about the `Verifier`.

Like in the case of `String` and `StringBuilder`.

You can see here an example of a builder implementation in C#: https://en.wikipedia.org/wiki/Builder_pattern

Reply:
I can make `VerifierBuilder` unaware of `Verifier` if I move `Build` to the extension if you don't want double direction dependency, but I don't see a practical advantage in copying all the references to another instance. When we already have an instance that holds the immutable source of truth. That would only lead to risk that some line will not be copied correctly and something will eventually get renamed out of sync.

Reply:
Shall we rename the thing to `VerifierParameters` or `VerifierConfiguration` then?